### PR TITLE
Architectural model

### PR DESCRIPTION
**Pull request by andreaslochbihler-da**
First draft of relaxing the ordering guarantees in the ledger model and incorporating transfers

@meiersi-da @gerolf-da @daravep @matthiasS-da: There's the question whether the write-up should start with a hypothetical totally ordered virtual shared ledger and relax the ordering to causality in party projections or start with causal order in party projection (for which we'd need some reads-from relation for contract keys) and then say that all the causal orders in all view projections can be combined into one hypothetical partially ordered virtual shared ledger. Please keep this question in mind when you review the write-up and voice your opinion.

TODO: Update the specification of the participant state API accordingly. 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto [hunk 1]
```diff
@@ -111,6 +113,87 @@ message ArchivedEvent {
...
+
+  // The arguments that have been used to create the contract.
+  // Required
+  Record create_arguments = 4;
+
+  // The parties that are notified of this event.
+  // These are the intersection of the parties specified in the `TransactionFilter`
+  // and the contract's signatories and observers whose view the contract enters.
+  // Required
+  repeated string witness_parties = 5;
```
Comment (oggy-) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto]: Why is this field useful?

File: docs/source/concepts/architectural-model.rst [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+#. Create a contract `c` with stakeholders `A` and `B` on source 2
+#. Exercise a non-consuming choice `ch1` on `c` on source 1.
+#. Exercise a non-consuming choice `ch2` on `c` on source 2.
+#. Exercise a consuming choice on `c` on source 1.
+
+Then, the transaction tree stream that `P1` provides for `A` contains five events around contract `c`: ``EnteredViewEvent``, ``ExercisedEvent``, ``LeftViewEvent``, ``EnteredViewEvent``, ``ArchivedEvent``.
+These five events can be grouped into between two or five transactions.
+For example, the first three and the last two could be batched into one transaction each.
+However, the ``LeftViewEvent`` cannot be combined with the subsequent ``EnteredViewEvent`` and they cannot be elided either.
```
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Then the contract stays in the participant's view. So there's no need for `LeftViewEvent` and `EnteredViewEvent` as the participant can be sure to receive all updates for the contract.

File: docs/source/concepts/architectural-model.rst [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+This section defines :ref:`causality <ledger-causality-party>` and illustrates the allowed and disallowed reorderings of commits by :ref:`examples <ledger-causality-examples>`.
+
+.. _ledger-causality-party:
+
+Causality
+=========
+
+Causality is defined on the commits in a party's projection of the virtual shared ledger.
+
+Definition »synchronizes with«
```
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: I've stolen the term `synchronizes with` from memory models for programming languages, see e.g. https://preshing.com/20130823/the-synchronizes-with-relation/ Happy to use a different name if something comes to your mind.

File: docs/source/concepts/architectural-model.rst [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    `act`\ :sub:`1` creates the contract `c` and `act`\ :sub:`2` uses the contract, i.e., `act`\ :sub:`2` is a **Fetch** or an **Exercise** on `c` and `A` is a stakeholder of `c` and an informee of `act`\ :sub:`2`.
+ 
+  * **Use-use:**
+    `act`\ :sub:`1` and `act`\ :sub:`2` both use a contract `c` and `A` is a stakeholder of `c` and an informee of both actions.
+
+  * **Contract keys**:
+    `act`\ :sub:`1` and `act`\ :sub:`2` both are actions on a key `k` and `A` is a maintainer of `k`.
+
+Definition »causal order«
+  A commit `C`\ :sub:`2` **causally depends on** another commit `C`\ :sub:`1` for party `A` if they contain actions `act`\ :sub:`2` and `act`\ :sub:`1` such that `act`\ :sub:`1` synchronizes with `act`\ :sub:`2` for party `A`.
```
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Argh. Good catch. I implicitly had in mind that the commits C1 and C2 come in this order on the transaction (tree) stream or on the virtual shared ledger. Will have to clarify.

File: docs/source/concepts/architectural-model.rst [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+.. _ledger-api-guarantees:
+  
+Ledger API ordering guarantees
+==============================
+
+Subject to the trust assumptions of the deployed ledger,
+if a party `A` is hosted on participant,
+the participant's transaction service output satisfy the following guarantees:
+
+#. The transaction trees on the transaction tree stream are precisely those that are in the participant's view for `A`.
```
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: True for the annotations. That's why the architectural model is a refinement of the ledger model. Maybe it's not yet clear from the text.

As for participants leaving and joining domains: I deliberately went for a static topology in this write-up. As soon as participants leave and join domains, we'll have to talk about party migration. And that's a mine field at the moment. Initially, I wanted to cover topology changes as well (and I believe that the view abstraction is suitable for what we do in Canton) but there's a lot of unconsensus on party migration and party management between the different ledgers that I consider it infeasible right now to make progress there.

File: docs/source/concepts/architectural-model.rst [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+   This excludes ``LeftViewEvent``\ s and ``EnteredViewEvent``\ s because they are not part of the virtual shared ledger.
+#. The flat transaction stream contains precisely the ``CreatedEvent``\ s and ``ArchivedEvent``\ s that are in the participant's view of the virtual shared ledger and which act on contracts where `A` is a stakeholder.
+#. The outputs on both streams are ordered such that their causal order for party `A` is the same as on the virtual shared ledger.
```
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Sounds reasonable.

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto (hunk 1 above)
Comment (andreaslochbihler-da) [reply ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto]: Just for symmetry with the other events. From what I can see from the documentation, the ledger API server uses the `witness_parties` field to determine whose event stream shall include the event.

File: docs/source/concepts/architectural-model.rst [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+The contract is reported in the active contract service and can be used by command submission.
+The ``EnteredViewEvent`` contains all the information in a ``CreatedEvent``;
+the only difference is that ``EnteredViewEvent``\ s may occur several times whereas there should be only at mose one ``CreatedEvent`` for each contract.
+
+These events are generated when the underlying commit protocol synchronizes between the different sources.
+This may happen as part of command submission or for other reasons, e.g., load balancing.
+It is guaranteed that the ``EnteredViewEvent`` precedes contract usage, subject to the trust assumptions of the underlying ledger.
+
+A contract may enter and leave the view of a participant several times.
+For example, suppose that party `B` submits the following commands on the given sources.
```
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Rephrased.

File: docs/source/concepts/architectural-model.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Clarified.

File: docs/source/concepts/architectural-model.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: And replaced `source` by `domain`.

File: docs/source/concepts/architectural-model.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Added a note to explain the asymmetry and that it does not matter for causal ordering in consistent ledgers.

File: docs/source/concepts/architectural-model.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Added the constraint that `C1` and `C2` are ordered by the virtual shared ledger.

File: docs/source/concepts/architectural-model.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Added a sentence that we assume that the virtual shared ledger annotates commits with domains.

File: docs/source/concepts/architectural-model.rst [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+the participant's transaction service output satisfy the following guarantees:
+
+#. The transaction trees on the transaction tree stream are precisely those that are in the participant's view for `A`.
+   This excludes ``LeftViewEvent``\ s and ``EnteredViewEvent``\ s because they are not part of the virtual shared ledger.
+#. The flat transaction stream contains precisely the ``CreatedEvent``\ s and ``ArchivedEvent``\ s that are in the participant's view of the virtual shared ledger and which act on contracts where `A` is a stakeholder.
+#. The outputs on both streams are ordered such that their causal order for party `A` is the same as on the virtual shared ledger.
+
+Similarly, the active contract service provides the set of contracts that are active at the returned offset according to the transaction service output.
+That is, all events from the transaction event stream are included in the provided set of contracts.
+In particular, the set of active contracts and contract keys can be kept up to date by processing the events from the flat transaction stream (or the transaction tree stream) without having to take events with earlier offsets into account.
```
[the remark this answers is not in the document]
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Rephrased. Please check again.

**Comment by andreaslochbihler-da** [comment]
@oggy- I've renamed `source` to `domain` and `participant node` to `endpoint`. I prefer to have the ledger topology section as a prerequisite, in particular given that the background section has been moved down; deployment now comes first.

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto [hunk 9]
```diff
@@ -111,6 +113,87 @@ message ArchivedEvent {
   repeated string witness_parties = 4;
 }
 
+// Records that a contract has entered the view of the participant.
+// The contract has been created on a ledger source on which the participant does not represent
```
Comment (bame-da) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto]:
```suggestion
// The contract has been created on a domain on which the participant does not represent
```

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto [hunk 10]
```diff
@@ -111,6 +113,87 @@ message ArchivedEvent {
...
+// all stakeholders of the contract or has previously left the view.
+// Used only for partitioned ledger topologies with multiple sources.
```
Comment (bame-da) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto]:
```suggestion
// Used only for partitioned ledger topologies with multiple domains.
```

File: docs/source/concepts/architectural-model.rst [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+Causality is defined on the commits in a party's projection of the virtual shared ledger.
+
+Definition »synchronizes with«
+  An action `act`\ :sub:`1` **synchronizes with** an action `act`\ :sub:`2` for party `A` if one of the following holds:
+
+  * **Create-use:**
+    `act`\ :sub:`1` creates the contract `c` and `act`\ :sub:`2` uses the contract, i.e., `act`\ :sub:`2` is a **Fetch** or an **Exercise** on `c` and `A` is a stakeholder of `c` and an informee of `act`\ :sub:`2`.
+ 
+  * **Use-use:**
+    `act`\ :sub:`1` and `act`\ :sub:`2` both use a contract `c` and `A` is a stakeholder of `c` and an informee of both actions.
```
Comment (andreaslochbihler-da) [inline docs/source/concepts/architectural-model.rst]: It is not clear that we want to enforce a causal order between two non-consuming usages of a contract. We might want to restrict this rule to the case where act2 consumes the contract.

File: docs/source/concepts/architectural-model.rst [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+  
+* The :ref:`ledger API services <ledger-api-services>`
+
+* The :ref:`DAML ledger topologies <daml-ledger-topologies>`
+
+   
+Multiple domains for the virtual shared ledger
+**********************************************
+
+In the DA ledger model, the virtual shared ledger consists of a :ref:`sequence of commits <ledger-structure>`.
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: Surely we need to change the phrase "sequence of commits". Otherwise you get into trouble trying to explain multi-domain topologies where the virtual shared ledger is no longer linearly ordered. Maybe we should switch to a "DAG of commits". Otherwise you'd have to say that each domain is a virtual shared ledger with a broader term for multi-domain setups. Virtual Shared Multi-Ledger? blergh

File: docs/source/concepts/architectural-model.rst [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+The virtual shared ledger is fed from three domains.
+The arrows indicate that `P1` hosts `A` on domains 1 and 3 and `P2` on domains 1 and 2 and `P3` hosts `B` on all three.
+
+.. figure:: ./images/multiple-domains.svg
+   :align: center
+   :name: multiple-domains
+
+   Virtual shared ledger topology with three domains
+
+A view on the virtual shared ledger
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
A View on the Virtual Shared Ledger
```

File: docs/source/concepts/architectural-model.rst [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+***********************************
+
+When an endpoint hosts a party only on a subset of the domains for the vitual shared ledger,
+then the endpoint's transaction and active contract services are derived only from commits from those domains.
+An endpoint therefore provides each hosted party with a **view** on their projection of the virtual shared ledger.
+The view consists of all the commits that originate from the domains on which the endpoint hosts the party.
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: This could be interpreted to give a guarantee that all Participants know all the history of a party on the domain. Maybe rephrase to make clear that you get ACS and new transactions for hosted parties.

File: docs/source/concepts/architectural-model.rst [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+This view determines what the party will see on the endpoint's transaction service and the active contract service.
+In practice, the endpoint assembles the commits from the different domains into the party's view.
+In the following, we assume that the virtual shared ledger annotates all commits with the originating domain.
+
+In the :ref:`above example <multiple-domains>`, when a transaction creates a contract with stakeholder `A` on domain 2, then this transaction is in `P2`\ 's view for `A`, but not in `P1`\ 's.
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: P1, P2, and P3 are not indicated in the diagram.

File: docs/source/concepts/architectural-model.rst [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+Accordingly, `P2` outputs the transaction and the create event on `A`\ 's event stream and reports the contract as active.
+In contrast, `P1` will neither output the event nor report the contract as active, as `P1` is not connected to domain 2.
+
+
+Contracts entering and leaving a view
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
Contracts Entering and Leaving a View
```

File: docs/source/concepts/architectural-model.rst [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+For interoperability, it is important that a transaction can use a contract whose creation comes from a different domain.
+In the :ref:`above example <multiple-domains>`, one transaction can create a contract with stakeholder `A` on domain 1 and another archives the contract on domain 2.
+Then endpoint `P1` outputs the ``CreatedEvent``, but not the ``ArchiveEvent`` on the transaction service.
+Conversely, when one transaction creates a contract with stakeholder `A` on domain 3 and another archives the contract on domain 1, then `P2` outputs the ``ArchivedEvent``, but not the ``CreateEvent``.
+
+To keep the transaction stream consistent, `P1` additionally outputs a ``LeftViewEvent`` on `A`\ 's transaction stream.
+This event signals that the endpoint no longer outputs events concerning this contract; in particular not when the contract is archived.
+The contract is no longer reported in the active contract service and cannot be used by command submissions.
+
+Conversely, `P2` outputs an ``EnteredViewEvent``\ s some time before the ``ArchivedEvent`` on the transaction stream.
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
Conversely, `P2` outputs an ``EnteredViewEvent`` some time before the ``ArchivedEvent`` on the transaction stream.
```

File: docs/source/concepts/architectural-model.rst [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+#. Exercise a non-consuming choice `ch2` on `c` on domain 2.
+#. Exercise a consuming choice on `c` on domain 1.
+
+Then, the transaction tree stream that `P1` provides for `A` contains five events around contract `c`: ``EnteredViewEvent``, ``ExercisedEvent``, ``LeftViewEvent``, ``EnteredViewEvent``, ``ArchivedEvent``.
+These five events can be grouped into between two or five transactions.
+For example, the first three and the last two could be batched into one transaction each.
+However, `P1` cannot combine the ``LeftViewEvent`` with the subsequent ``EnteredViewEvent`` and `P1` must not elide them either.
+This is because their presence indicates that `P1`\ 's view may miss some events in between; in this example, exercising the choice `ch2`.
+
+The flat transaction stream by `P1` shows omits the non-consuming exercise choices.
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
The flat transaction stream by `P1` omits the non-consuming exercise choices.
```

File: docs/source/concepts/architectural-model.rst [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+It nevertheless contains the three view change events ``EnteredViewEvent``, ``LeftViewEvent``, and ``EnteredViewEvent`` before the ``ArchivedEvent``.
+This is because the endpoint cannot know at the ``LeftViewEvent`` that there will be another ``EnteredViewEvent``.
+
+In contrast, `P2` need not output the ``EnteredViewEvent``\ s and ``LeftViewEvent``\ s at all in this example because `P2` hosts `A` on both domains.
+
+
+Relaxing the total ordering
+***************************
+
+The virtual shared ledger and the transaction service put all transactions into a linear sequence.
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: Do they? How?

File: docs/source/concepts/architectural-model.rst [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+
+Causality is defined on the commits in a party's projection of the virtual shared ledger.
+
+Definition »synchronizes with«
+  An action `act`\ :sub:`1` **synchronizes with** an action `act`\ :sub:`2` for party `A` if one of the following holds:
+
+  * **Create-use:**
+    `act`\ :sub:`1` creates the contract `c` and `act`\ :sub:`2` uses the contract, i.e., `act`\ :sub:`2` is a **Fetch** or an **Exercise** on `c` and `A` is a stakeholder of `c` and an informee of `act`\ :sub:`2`.
+ 
+  * **Use-use:**
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: Does "use" include non-consuming actions? Do we really want two fetches to synchronize?

File: docs/source/concepts/architectural-model.rst [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+    `act`\ :sub:`1` and `act`\ :sub:`2` both use a contract `c` and `A` is a stakeholder of `c` and an informee of both actions.
+
+  * **Contract keys**:
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: Again, does "action" here include lookups? Do we really want to lookups to synchronize?

File: docs/source/concepts/architectural-model.rst [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+    `act`\ :sub:`1` and `act`\ :sub:`2` both are actions on a key `k` and `A` is a maintainer of `k`.
+
+Definition »causal order«
+  A commit `C`\ :sub:`2` **causally depends on** another commit `C`\ :sub:`1` for party `A` if all of the following holds:
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
  A commit `C`\ :sub:`2` **causally depends on** another commit `C`\ :sub:`1` for party `A` if all of the following holds:
  
```

File: docs/source/concepts/architectural-model.rst [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+  * `C`\ :sub:`1` precedes `C`\ :sub:`2` on the virtual shared ledger.
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: This seems very circular to me. To me, you are defining the partial order of the VSL here, the ordering being causal dependency. Yet you use the order of the VSL to define it.

File: docs/source/concepts/architectural-model.rst [hunk 24]
```diff
@@ -0,0 +1,284 @@
...
+
+.. note::
+   The synchronizes with relationship is not symmetric.
+   If `act`\ :sub:`1` uses a contract and `act`\ :sub:`2` creates it, then `act`\ :sub:`1` does not synchronize with `act` :sub:`2`.
+   This is in line with how the concept is used in programming languages: writes to a memory location synchronize with reads, but reads do not synchronize with writes.
+   This asymmetry does not matter for the causal ordering in a consistent ledger as **Create**\ s precede usages on such a ledger.
+  
+.. _ledger-api-guarantees:
+  
+Ledger API ordering guarantees
```
Comment (bame-da) [inline docs/source/concepts/architectural-model.rst]: The way you've set this doc up makes this last part, which is what people actually care about, very complicated.

- VSLs are a DAG
- A Party projection is a subgraph (and thus a DAG), sub-ledger, subVSL
- A Party projection is a subgraph (and thus a DAG), sub-ledger, subVSL
- On any such graph you can define the transitive closure of the ordering, which is causality order
- Ledger API guarantees to return transactions according to causality order for the transaction graph available for thre party on the Participant Node

File: docs/source/concepts/architectural-model.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (bame-da) [reply docs/source/concepts/architectural-model.rst]: I suggest "directly depends on" vs "causally depends on"

File: docs/source/concepts/architectural-model.rst (hunk 11 above)
Comment (bame-da) [reply docs/source/concepts/architectural-model.rst]: You beat me to it, but I took a long time to process this PR. I don't think we should enforce causal order between non-consuming uses.

File: docs/source/concepts/architectural-model.rst (hunk 21 above)
Comment (andreaslochbihler-da) [reply docs/source/concepts/architectural-model.rst]: Yes if it is a `NoSuchKey` action (the ledger model models a positive `lookupByKey` as a `Fetch` of the corresponding contract).

This discussion is the analogous as for `use-use` of contracts. If we decide that non-consuming uses do not synchronize with each other, then there's no point in having `NoSuchKey` actions synchronize with each other either. And vice versa.

Comment (daravep) [reply docs/source/concepts/architectural-model.rst]: Hmm ... but wouldn't you just have to synchronise some key actions? I.e. you don't need to synchronise two reads, you only need to synchronise if one action is a write action.

File: docs/source/concepts/architectural-model.rst [hunk 25]
```diff
@@ -0,0 +1,284 @@
...
+==============================
+
+Subject to the trust assumptions of the deployed ledger,
+if a party `A` is hosted on an endpoint,
+the endpoint's transaction service output for `A` satisfy the following guarantees:
+
+#. The transaction trees on the transaction tree stream are precisely those that are in the endpoint's view for `A`.
+   This excludes ``LeftViewEvent``\ s and ``EnteredViewEvent``\ s because they are not part of the virtual shared ledger.
```
Comment (daravep) [inline docs/source/concepts/architectural-model.rst]: Ah, so that's actually interesting. By calling this out, you distinguish between the guarantees that the VSL gives you versus the guarantees that the VIEW on the VSL gives you.

File: docs/source/concepts/architectural-model.rst [hunk 26]
```diff
@@ -0,0 +1,284 @@
...
+.. image:: ./ledger-model/images/divulgence-for-disclosure-counteroffer.svg
+   :align: center
+   :width: 100%
+
+#. *When two parties are stakeholders on the same contract,
+   then they see creation and archival in the same order.*
+   
+   In the example, the creation of the `CounterOffer A P Bank` synchronizes with the painter exercising the consuming choice on the `CounterOffer` for both Alice and the painter by the **create-use** rule, since they both are stakeholders on the contract.
+   Accordingly, Alice will see the creation before the archival on her transaction stream and so will the painter.
+   This does not depend on whether they are hosted on the same endpoinnt or whether the **Create** comes from the same domain as the **Exercise**.
```
Comment (daravep) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
   This does not depend on whether they are hosted on the same endpoint or whether the **Create** originates from the same domain as the **Exercise**.
```

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto [hunk 27]
```diff
@@ -111,6 +113,87 @@ message ArchivedEvent {
...
+  // Required
+  Identifier template_id = 3;
+
+  // The key of the contract, if defined.
+  // Optional
+  Value contract_key = 7;
+
+  // The arguments that have been used to create the contract.
+  // Required
+  Record create_arguments = 4;
```
Comment (daravep) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto]: It might be premature to talk about this already now, but I would assume that at some point, we might want to avoid sending the payload several times and therefore split the payload from status updates and only send the payload once.

File: ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto [hunk 28]
```diff
@@ -26,6 +26,8 @@ message Event {
   oneof event {
     CreatedEvent created = 1;
     ArchivedEvent archived = 3;
+    EnteredViewEvent enteredView = 4;
+    LeftViewEvent leftView = 5;
```
Comment (daravep) [inline ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/event.proto]: Given that you clean up here, it might make sense to also add the suggested changes to distinguish between actual created events (the ones where we will see an archive if it is archived) and just disclosed create events.

File: docs/source/concepts/architectural-model.rst [hunk 29]
```diff
@@ -0,0 +1,284 @@
...
+The components in this diagram are the following:
+
+* There is a set of **domains** for the virtual shared ledger.
+  Each domain contributes commits to the virtual shared ledger.
+
+* There is a set of **endpoints**.
+  Each endpoint is connected to a subset of the **domains**.
+
+* Endpoints host parties on a subset of the domains they are connected to.
+  An endpoint's services provide a party with projected commits from the domains on which it hosts the party.
```
[the remark this answers is not in the document]
Comment (matthiasS-da) [reply docs/source/concepts/architectural-model.rst]: MINOR: I would replace "provide ... with" by "inform ... about"

File: docs/source/concepts/architectural-model.rst [hunk 30]
```diff
@@ -0,0 +1,284 @@
...
+Replication and concensus ensures that all writer nodes agree on the contents of the virtual shared ledger.
+Conceptually, they form a single domain even though endpoints may connect only to a subset of the writer nodes.
+
+Multiple domains appear with interoperability in :ref:`partitioned ledger topologies <partitioned-topologies>`.
+Interoperability makes it possible that an atomic transaction uses the output contracts of transactions that have been recorded to different deployed ledgers.
+For example, when an organization initially deploys two workflows to two ledgers, it can later compose those workflows into a larger workflow that spans the deployed ledger.
+In this setup, each deployed ledger constitutes a domain for the virtual shared ledger.
+Multiple domains can also be used to improve horizontal scaling of the atomic commit protocol.
+Endpoints participate only in the commit protocols of the domains they are connected to and the commit protocols of different domains can often run in parallel.
+The next diagram shows a such situation where party `A` is hosted on two endpoints `P1` and `P2` and party `B` is hosted on `P3`.
```
Comment (matthiasS-da) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
The next diagram shows such a situation where party `A` is hosted on two endpoints `P1` and `P2` and party `B` is hosted on `P3`.
```

File: docs/source/concepts/architectural-model.rst [hunk 31]
```diff
@@ -0,0 +1,284 @@
...
+.. figure:: ./images/multiple-domains.svg
+   :align: center
+   :name: multiple-domains
+
+   Virtual shared ledger topology with three domains
+
+A view on the virtual shared ledger
+***********************************
+
+When an endpoint hosts a party only on a subset of the domains for the vitual shared ledger,
```
Comment (matthiasS-da) [inline docs/source/concepts/architectural-model.rst]:
```suggestion
When an endpoint hosts a party only on a subset of the domains for the virtual shared ledger,
```

File: docs/source/concepts/architectural-model.rst [hunk 32]
```diff
@@ -0,0 +1,284 @@
...
+Interoperability makes it possible that an atomic transaction uses the output contracts of transactions that have been recorded to different deployed ledgers.
+For example, when an organization initially deploys two workflows to two ledgers, it can later compose those workflows into a larger workflow that spans the deployed ledger.
+In this setup, each deployed ledger constitutes a domain for the virtual shared ledger.
+Multiple domains can also be used to improve horizontal scaling of the atomic commit protocol.
+Endpoints participate only in the commit protocols of the domains they are connected to and the commit protocols of different domains can often run in parallel.
+The next diagram shows a such situation where party `A` is hosted on two endpoints `P1` and `P2` and party `B` is hosted on `P3`.
+The virtual shared ledger is fed from three domains.
+The arrows indicate that `P1` hosts `A` on domains 1 and 3 and `P2` on domains 1 and 2 and `P3` hosts `B` on all three.
+
+.. figure:: ./images/multiple-domains.svg
```
Comment (matthiasS-da) [inline docs/source/concepts/architectural-model.rst]: Please label the endpoints in the picture with `P1`, `P2`, `P3`.

File: docs/source/concepts/architectural-model.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (matthiasS-da) [reply docs/source/concepts/architectural-model.rst]: I fully agree with @bame-da. The term "synchronizes with" sounds odd to me.
For direct dependencies (btw actions / commits) use "directly depends on".
For the transitive closure use "causally depends on".

File: docs/source/concepts/architectural-model.rst [hunk 33]
```diff
@@ -0,0 +1,284 @@
...
+   :name: multiple-domains
+
+   Virtual shared ledger topology with three domains
+
+A view on the virtual shared ledger
+***********************************
+
+When an endpoint hosts a party only on a subset of the domains for the vitual shared ledger,
+then the endpoint's transaction and active contract services are derived only from commits from those domains.
+An endpoint therefore provides each hosted party with a **view** on their projection of the virtual shared ledger.
```
Comment (matthiasS-da) [inline docs/source/concepts/architectural-model.rst]: Please don't define terms on the fly. Further down, I wanted to jump back to recall the definition, but I had a hard time finding it. So please make it a proper definition.

File: docs/source/concepts/architectural-model.rst (hunk 14 above)
Comment (matthiasS-da) [inline docs/source/concepts/architectural-model.rst]: The terms "view" and "projection" seem to be very similar. Therefore, I disagree with introducing a new term. We already have "projection of a party". I would extend this with:
- "projection of a participant" (all the participant sees)
- "projection of a party and participant" (all the party sees through the participant)
